### PR TITLE
Variable histogram

### PR DESCRIPTION
**Pull request by bilgili**

### CODE REVIEW

**Comment by bilgili** [comment]
retest this please

File: livre/core/data/Histogram.h [hunk 1]
```diff
@@ -82,12 +82,23 @@ class Histogram : public co::Distributable< ::lexis::render::Histogram >
...
+    /**
      * Computes the ratio of the value at a given index.
      * @param index the index of the histogram value
      * @return the ratio at given index. If histogram is empty or index exceeds
      * the histogram bin count, returns 0.0.
      */
     LIVRECORE_API double getRatio( size_t index ) const;
+
+    /**
+      * Clears the histogram and sets the number of bins to newSize
```
Comment (hernando) [inline livre/core/data/Histogram.h]: Changes the number of bins to newSize and clears the histogram.

File: livre/eq/Config.cpp [hunk 2]
```diff
@@ -94,14 +100,27 @@ class Config::Impl
...
     {
-      // If we get a very old frame skip it
+        // If we get a very old frame skip it
         if( !histogramQueue.empty() && currentId < histogramQueue.back().id )
             return;
 
+        // Always enlarge the range
+        dataSourceRange[ 0 ] =
+                histogram.getMin() < dataSourceRange[ 0 ] ? histogram.getMin() :
+                                                            dataSourceRange[ 0 ];
```
Comment (hernando) [inline livre/eq/Config.cpp]: The code may look better if you use an if statement.

File: livre/eq/Channel.cpp [hunk 3]
```diff
@@ -291,6 +292,7 @@ struct Channel::Impl
         renderPipeline.render( pipe->getFrameData()->getVRParameters(),
                                _frameInfo,
                                {{ _drawRange.start, _drawRange.end }},
+                               pipe->getFrameData()->getVolumeSettings().getDataSourceRange(),
```
Comment (hernando) [inline livre/eq/Channel.cpp]: I think I already mentioned in other code review that this function has too many parameters. Can you create a struct to pass them.

File: livre/eq/Config.cpp [hunk 4]
```diff
@@ -111,8 +130,16 @@ class Config::Impl
...
             if( currentId == data.id )
             {
-                dataMerged = true;
-                data += viewHistogram;
+                try
+                {
+                    data += viewHistogram;
+                    dataMerged = true;
+                }
+                catch( std::runtime_error& )
```
Comment (hernando) [inline livre/eq/Config.cpp]: Add a comment that explains when this exception is thrown.

File: livre/eq/Config.cpp [hunk 5]
```diff
@@ -94,14 +100,27 @@ class Config::Impl
...
     {}
 
     void gatherHistogram( const Histogram& histogram, const float area, const uint32_t currentId )
     {
-      // If we get a very old frame skip it
+        // If we get a very old frame skip it
         if( !histogramQueue.empty() && currentId < histogramQueue.back().id )
             return;
 
+        // Always enlarge the range
```
Comment (hernando) [inline livre/eq/Config.cpp]: Extend the global histogram range if needed.

File: livre/eq/Config.cpp [hunk 6]
```diff
@@ -94,14 +100,27 @@ class Config::Impl
...
+        dataSourceRange[ 0 ] =
+                histogram.getMin() < dataSourceRange[ 0 ] ? histogram.getMin() :
+                                                            dataSourceRange[ 0 ];
+
+        dataSourceRange[ 1 ] =
+                histogram.getMax() > dataSourceRange[ 1 ] ? histogram.getMax() :
+                                                            dataSourceRange[ 1 ];
+
+        config->getFrameData().getVolumeSettings().setDataSourceRange( dataSourceRange );
```
Comment (hernando) [inline livre/eq/Config.cpp]: Add coment: "Updating the range that clients must use to create the partial histograms."

File: livre/eq/Config.cpp [hunk 7]
```diff
@@ -163,6 +190,7 @@ class Config::Impl
     bool redraw;
     Vector2ui dataFrameRange;
     ViewHistogramQueue histogramQueue;
+    Vector2f dataSourceRange;
```
Comment (hernando) [inline livre/eq/Config.cpp]: histogramRange?

File: livre/eq/Config.cpp [hunk 8]
```diff
@@ -94,14 +100,27 @@ class Config::Impl
         , eventMapper( EventHandlerFactoryPtr( new EqEventHandlerFactory ))
         , redraw( true )
         , dataFrameRange( INVALID_FRAME_RANGE )
+        , dataSourceRange( std::numeric_limits<uint8_t>::min(), // This will change with the new TF
+                           std::numeric_limits<uint8_t>::max( ))
```
Comment (hernando) [inline livre/eq/Config.cpp]: This means the default range is [0, 255.0]? Why not writing it explicitly instead of using std::numeric_limits, it's more obvious that way.

File: livre/eq/Channel.cpp (hunk 3 above)
Comment (bilgili) [inline livre/eq/Channel.cpp]: :) I was afraid somebody will ask for it ( You are not the first, but you are right, I am postponing the in-evidable )

File: livre/eq/Config.cpp (hunk 8 above)
Comment (hernando) [inline livre/eq/Config.cpp]: You can also add that this range assumes that data sources are 8 bit by default.

File: livre/core/data/Histogram.h (hunk 1 above)
Comment (bilgili) [inline livre/core/data/Histogram.h]: Internally does what I say :) I don't know which, but ok.

File: livre/eq/Config.cpp (hunk 7 above)
Comment (bilgili) [inline livre/eq/Config.cpp]: It is really the dataSource range, histogram adapts to it.

File: livre/eq/Config.cpp [hunk 9]
```diff
@@ -94,14 +100,27 @@ class Config::Impl
...
+
+        dataSourceRange[ 1 ] =
+                histogram.getMax() > dataSourceRange[ 1 ] ? histogram.getMax() :
+                                                            dataSourceRange[ 1 ];
+
+        config->getFrameData().getVolumeSettings().setDataSourceRange( dataSourceRange );
+
         const ViewHistogram viewHistogram( histogram, area, currentId );
         auto it = histogramQueue.begin();
         while( it != histogramQueue.end( ))
```
Comment (hernando) [inline livre/eq/Config.cpp]: Use a for? You are always increasing it at the end of the iteration except if you break or return.

File: livre/eq/settings/VolumeSettings.cpp [hunk 10]
```diff
@@ -33,31 +33,44 @@ VolumeSettings::VolumeSettings()
...
+    _dataSourceRange = range;
+    setDirty( DIRTY_ALL );
+}
+
+const vmml::Vector2f& VolumeSettings::getDataSourceRange() const
+{
+    return _dataSourceRange;
+}
+
+void VolumeSettings::setURI( const std::string& uri )
```
Comment (hernando) [inline livre/eq/settings/VolumeSettings.cpp]: Keep setURI close to getURI as it was to reduce the diff.

File: livre/eq/settings/VolumeSettings.h [hunk 11]
```diff
@@ -35,33 +36,37 @@ class VolumeSettings : public co::Serializable
...
     /**
-     * @return Returns the uri of the source data.
+     * Sets the accumulated data range for the volume.
+     * @param range of the data source.
      */
-    const std::string& getURI( ) const;
+    void setDataSourceRange( const Vector2f& range );
+
+    /** @return Returns the accumulated data range for the volume. */
+    const Vector2f& getDataSourceRange() const;
```
Comment (hernando) [inline livre/eq/settings/VolumeSettings.h]: I feel this should be called differently, like getHistogramRange or getGlobalDataRange, in any case remove Source, I don't think it's needed.

File: livre/lib/cache/HistogramObject.cpp [hunk 12]
```diff
@@ -40,14 +41,13 @@ void binData( const SRC_TYPE* rawData,
               const size_t compCount,
               const uint64_t scaleFactor )
```
Comment (hernando) [inline livre/lib/cache/HistogramObject.cpp]: I can suggest you a better refactoring of this code to reduce copy&paste.

File: livre/eq/settings/VolumeSettings.h (hunk 11 above)
Comment (bilgili) [inline livre/eq/settings/VolumeSettings.h]: Unfortunately, it mixes with equalizer Sort Last "datarange". That is why I called that one renderdatarange, and this one datasource range.

File: livre/lib/cache/HistogramObject.cpp (hunk 12 above)
Comment (bilgili) [inline livre/lib/cache/HistogramObject.cpp]: I will keep it this way. I don't feel refactoring will be cleaner.

File: livre/core/data/Histogram.cpp [hunk 13]
```diff
@@ -97,35 +107,49 @@ bool Histogram::isEmpty() const
...
 
     return sum;
 }
 
+Vector2f Histogram::getRange() const
+{
+    return Vector2f( getMin(), getMax( ));
+}
+
 double Histogram::getRatio( const size_t index ) const
```
Comment (hernando) [inline livre/core/data/Histogram.cpp]: Not that is important, but every time you call this function you loop over all the histogram to compute the sum of all counts. Do you need to get the ratio of a single index or in general you need all of them (for example for the GUI), because for the latter case you could have a function that returns a std::vector<float> with the normalized values and also remove the loop that the caller needs.

File: livre/lib/pipeline/HistogramFilter.cpp [hunk 14]
```diff
@@ -80,6 +80,7 @@ struct HistogramFilter::Impl
         const auto& frustums = input.get< Frustum >( "Frustum" );
         const auto& viewports = input.get< Viewport >( "RelativeViewport" );
 
+        auto dataSourceRange = input.get< Vector2f >( "DataSourceRange" ).front();
         const auto& frustum = frustums.front();
```
Comment (hernando) [inline livre/lib/pipeline/HistogramFilter.cpp]: Why don't you do the same thing for frustum and viewport, I mean: 

    const auto& frustum = input.get< Frustum >( "Frustum" ).front();

File: livre/lib/pipeline/HistogramFilter.cpp [hunk 15]
```diff
@@ -88,26 +89,48 @@ struct HistogramFilter::Impl
...
                         _histogramCache.load< HistogramObject >( cacheId,
                                                                  _dataCache,
-                                                                 _dataSource );
+                                                                 _dataSource,
+                                                                 dataSourceRange );
                 if( !histCacheObject )
                     continue;
 
-                ConstHistogramObjectPtr histogramObj =
-                        std::static_pointer_cast< const HistogramObject >(
```
Comment (hernando) [inline livre/lib/pipeline/HistogramFilter.cpp]: Why did you need this cast before if load is already templated?

File: livre/lib/cache/HistogramObject.cpp [hunk 16]
```diff
@@ -58,10 +58,64 @@ void binData( const SRC_TYPE* rawData,
                                          compCount * j * dataBlockSize.z() +
                                          compCount * k + c;
                     const double data = rawData[ index ];
-                    const size_t binIndex =
-                            std::is_unsigned< SRC_TYPE >::value ?
-                            ( data - minVal ) / range * ( binCount - 1 ):
-                            (( data / range ) + 0.5f ) * ( binCount - 1 );
+                    const size_t binIndex =  histogram.getMin() + ( data - histogram.getMin( )) /
```
Comment (hernando) [inline livre/lib/cache/HistogramObject.cpp]: Add an assert to ensure than range is divisible by binCount and try not to use floating point operations for integral types. Assuming that range is divisible by binCount that should be fairly easy.
Besides that, the code is not correct for signed types, e.g. int8 and a value of -100 produce this:
-128 + ( -100 - -127) / 255 * 255 = -128 + 27/255*255 = -128 + 27 =  -101 which when converted to size_t will produce an interesting result.

File: livre/eq/Config.cpp (hunk 9 above)
Comment (bilgili) [inline livre/eq/Config.cpp]: You can not use erase, emplace if you use if. The compiler complains AFAIK. wont change

File: livre/lib/pipeline/HistogramFilter.cpp (hunk 15 above)
Comment (bilgili) [inline livre/lib/pipeline/HistogramFilter.cpp]: It was not templated before ?

Comment (hernando) [inline livre/lib/pipeline/HistogramFilter.cpp]: OK, so it was a cleanup that could have been done before this PR.

File: livre/eq/Channel.cpp [hunk 17]
```diff
@@ -289,15 +289,16 @@ struct Channel::Impl
...
-                               pipe->getFrameData()->getVolumeSettings().getDataSourceRange(),
-                               PixelViewport( pixVp.x, pixVp.y, pixVp.w, pixVp.h ),
-                               Viewport( vp.x, vp.y, vp.w, vp.h ),
+        renderPipeline.render( { pipe->getFrameData()->getVRParameters(),
+                                 _frameInfo,
+                                 {{ _drawRange.start, _drawRange.end }},
+                                 pipe->getFrameData()->getVolumeSettings().getDataSourceRange(),
+                                 PixelViewport( pixVp.x, pixVp.y, pixVp.w, pixVp.h ),
+                                 Viewport( vp.x, vp.y, vp.w, vp.h ),
+                                 pipe->getFrameData()->getRenderSettings().getClipPlanes(),
```
Comment (hernando) [inline livre/eq/Channel.cpp]: Is the extra comma at the end of the line allowed?

File: livre/core/data/Histogram.h [hunk 18]
```diff
@@ -34,7 +34,7 @@ class Histogram : public co::Distributable< ::lexis::render::Histogram >
 {
 public:
     /**
-     * Constructor. Initializes histogram with 0 values
+     * Constructor.
```
Comment (hernando) [inline livre/core/data/Histogram.h]: Specify the min, max and number of bins of a default constructed histogram?

File: livre/eq/settings/VolumeSettings.h (hunk 11 above)
Comment (hernando) [inline livre/eq/settings/VolumeSettings.h]: I understand that's the case in other places, but to me there's no ambiguity possible in this class (using your argument for a method called size we should have a vectorSize, listSize, ...).
You can also call it ValueRange. That 's even clearer than DataSourceRange in my opinion.

File: livre/lib/cache/HistogramObject.cpp [hunk 19]
```diff
@@ -40,14 +41,48 @@ void binData( const SRC_TYPE* rawData,
...
+        const Vector3ui dataBlockSize = blockSize + padding * 2;
+        for( size_t i = padding.x(); i < dataBlockSize.x() - padding.x(); ++i )
+            for( size_t j = padding.y(); j < dataBlockSize.y() - padding.y(); ++j )
+                for( size_t k = padding.z(); k < dataBlockSize.z() - padding.z(); ++k )
+                    for( size_t c = 0; c < compCount; ++c )
+                    {
+                        const size_t index = compCount * i * dataBlockSize.y() * dataBlockSize.z() +
+                                             compCount * j * dataBlockSize.z() +
+                                             compCount * k + c;
+                        const double data = rawData[ index ];
```
Comment (hernando) [inline livre/lib/cache/HistogramObject.cpp]: This doesn't work with volumes of complex numbers :-P

File: livre/lib/cache/HistogramObject.cpp [hunk 20]
```diff
@@ -40,14 +41,48 @@ void binData( const SRC_TYPE* rawData,
...
+                            maxVal = data;
+                    }
+
+        histogram.setMin( minVal );
+        histogram.setMax( maxVal );
+
+        if(( maxVal - minVal ) == 0.0f )
+        {
+            histogram.getBins().clear();
+            const size_t bins = ( dataBlockSize - padding ).product() * scaleFactor * compCount;
```
Comment (hernando) [inline livre/lib/cache/HistogramObject.cpp]: What is this scaleFactor? In line 98 below you add, it but here you are multiplying.

File: livre/lib/cache/HistogramObject.cpp [hunk 21]
```diff
@@ -57,11 +92,9 @@ void binData( const SRC_TYPE* rawData,
...
                                          compCount * j * dataBlockSize.z() +
                                          compCount * k + c;
-                    const double data = rawData[ index ];
-                    const size_t binIndex =
-                            std::is_unsigned< SRC_TYPE >::value ?
-                            ( data - minVal ) / range * ( binCount - 1 ):
-                            (( data / range ) + 0.5f ) * ( binCount - 1 );
+
+                    const size_t binIndex = ( rawData[ index ] - histogram.getMin( )) /
+                                            range * ( binCount - 1 );
```
Comment (hernando) [inline livre/lib/cache/HistogramObject.cpp]: This operation is not correct for integral types. Imagine you have a 16 bit data range and you want to map it to a histogram of 1024 bins. The mapping of values is [n, 64 * n) -> n, so 64 goes into bin 1, right? However if you compute 64 / 65535.0 * 1023.0 the result is 0.999... instead of 1. 

A correct solution is ( rawData[ index ] - histogram.getMin( )) / (range / binCount);

File: livre/eq/Channel.cpp (hunk 17 above)
Comment (bilgili) [inline livre/eq/Channel.cpp]: yep but I should remove it

File: livre/lib/pipeline/HistogramFilter.cpp (hunk 15 above)
Comment (bilgili) [inline livre/lib/pipeline/HistogramFilter.cpp]: yep

File: livre/lib/cache/HistogramObject.cpp (hunk 19 above)
Comment (bilgili) [inline livre/lib/cache/HistogramObject.cpp]: and 6 dimensional cartesian system as well :)

File: livre/lib/cache/HistogramObject.cpp (hunk 20 above)
Comment (bilgili) [inline livre/lib/cache/HistogramObject.cpp]: depending on the lod, number of voxels per data point.

File: livre/core/data/Histogram.cpp (hunk 13 above)
Comment (bilgili) [inline livre/core/data/Histogram.cpp]: It is only needed when hovering over the indices and it is not so expensive.

File: livre/lib/cache/HistogramObject.cpp (hunk 20 above)
Comment (hernando) [inline livre/lib/cache/HistogramObject.cpp]: OK, I see.

File: livre/lib/cache/HistogramObject.cpp [hunk 22]
```diff
@@ -93,8 +93,27 @@ void binData( const SRC_TYPE* rawData,
                                          compCount * j * dataBlockSize.z() +
                                          compCount * k + c;
 
-                    const size_t binIndex = ( rawData[ index ] - histogram.getMin( )) /
-                                            range * ( binCount - 1 );
+                    size_t binIndex = 0;
+                    if( std::is_integral< SRC_TYPE >::value )
+                    {
+                        const size_t range =
+                                std::lround( histogram.getMax() - histogram.getMin( )) + 1u;
```
Comment (hernando) [inline livre/lib/cache/HistogramObject.cpp]: You don't need lround for integers.

Comment (hernando) [inline livre/lib/cache/HistogramObject.cpp]: I'd try to move the range calculation outside of the quadruple loop. You can use some template tricks to solve the problem of the different datatypes.

Comment (hernando) [inline livre/lib/cache/HistogramObject.cpp]: But if you think it won't make a big difference leave it like this.

**Comment by hernando** [comment]
+1, squash the 3 commits before merging.

File: livre/lib/cache/HistogramObject.cpp (hunk 22 above)
Comment (bilgili) [inline livre/lib/cache/HistogramObject.cpp]: Wont make :)
